src/components/Tables/Filters.tsx

```diff
@@ -0,0 +1,100 @@
+import React from "react";
+import { styled } from "styled-components";
+import { TableHeaders } from "@/components/Tables/Table";
+
+export interface Filter<Data, State> {
+    shouldFilter(data: Data, state: State): boolean;
+    filterComponent(state: State, setState: (state: State) => void): React.ReactElement;
+    state: State;
+    initialState: State;
+    clear(state: State, setState: (state: State) => void): void;
+}
+
+export interface KeyedFilter<Data, N extends keyof Data, State> extends Filter<Data, State> {
+    key: N;
+    label: string;
+}
+
+const keyedFilter = <Data, N extends keyof Data, State>(
+    key: N,
+    label: string,
+    filter: Filter<Data, State>
+): KeyedFilter<Data, N, State> => {
+    return {
+        key,
+        label,
+        ...filter,
+    };
+};
+
+export const headerLabelFromKey = <Data, N extends keyof Data>(
+    headers: TableHeaders<Data>,
+    key: N
+): string => {
+    return headers.find(([headerKey]) => headerKey === key)![1];
+};
+
+const StyledFilterBar = styled.input`
+    font-size: 14px;
+    width: 15rem;
+    overflow: visible;
+    box-shadow: none;
+    padding: 4px 12px 4px 12px;
+
+    &:focus {
+        outline: none;
+    }
+`;
+
+interface TextFilterProps<Data, N extends keyof Data> {
+    key: N;
+    label: string;
+    caseSensitive?: boolean;
+    initialValue?: string;
+}
+
+const defaultToString = (value: any): string => {
+    if (typeof value === "string") {
+        return value;
+    }
+
+    return JSON.stringify(value);
+};
+
+export const textFilter = <Data, N extends keyof Data>({
+    key,
+    label,
+    caseSensitive = false,
+    initialValue = "",
+}: TextFilterProps<Data, N>): KeyedFilter<Data, N, string> => {
+    return keyedFilter(key, label, {
+        state: initialValue,
+        initialState: initialValue,
+        shouldFilter: (data, state) => {
+            let string = defaultToString(data[key]);
+
+            if (!caseSensitive) {
+                string = string.toLowerCase();
+                state = state.toLowerCase();
+            }
+
+            return !string.includes(state);
```

Comment:
Do you want it to return True if `string === ""`? (just checking)

Reply:
If using this check, I think simply write `if (!string)` will do?

Reply:
probably not? not sure what expected behaviour would be here